### PR TITLE
[8.0][IMP][contract_payment_mode] Copy partner payment mode to contracts when installing

### PR DESCRIPTION
**Pull request by antespi**
This improvement copies payment mode (configured in customer) to contracts when installing the addon.

### CODE REVIEW

File: contract_payment_mode/__init__.py [hunk 1]
```diff
@@ -1,2 +1,27 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
 from . import models
+from openerp import api, SUPERUSER_ID
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+def copy_payment_mode(cr, registry):
```
Comment (pedrobaeza) [inline contract_payment_mode/__init__.py]: Please put this into a hooks.py file according conventions: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#installation-hooks

File: contract_payment_mode/__init__.py [hunk 2]
```diff
@@ -1,2 +1,27 @@
...
+        if contracts:
+            _logger.info('Setting payment mode: %d contracts' %
+                         len(contracts))
+        for contract in contracts:
+            payment_mode = contract.partner_id.customer_payment_mode
+            if payment_mode:
+                contract.write({
+                    'payment_mode_id': payment_mode.id,
+                })
+        _logger.info('Setting payment mode: Done')
```
Comment (pedrobaeza) [inline contract_payment_mode/__init__.py]: This is always log independently if there are contracts or not

File: contract_payment_mode/__init__.py (hunk 1 above)
Comment (antespi) [inline contract_payment_mode/__init__.py]: OK, this is new for me. I've copy-pasted from other OCA addon ... Good to know, thanks

File: contract_payment_mode/__init__.py [hunk 3]
```diff
@@ -1,2 +1,27 @@
...
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        m_contract = env['account.analytic.account']
+        contracts = m_contract.search([('type', '=', 'contract')])
+        if contracts:
+            _logger.info('Setting payment mode: %d contracts' %
+                         len(contracts))
+        for contract in contracts:
+            payment_mode = contract.partner_id.customer_payment_mode
+            if payment_mode:
+                contract.write({
```
Comment (pedrobaeza) [inline contract_payment_mode/__init__.py]: You can write the compact form: `contract.payment_mode_id = payment_mode.id`

**Comment by antespi** [comment]
@pedrobaeza fixes done, thanks.

File: contract_payment_mode/__init__.py (hunk 2 above)
Comment (antespi) [inline contract_payment_mode/__init__.py]: That's the point :smile:

Comment (pedrobaeza) [inline contract_payment_mode/__init__.py]: But the other is not logged always.

**Comment by sergio-teruel** [comment]
I tested in runbot and :+1:

File: contract_payment_mode/hooks.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    with api.Environment.manage():
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        m_contract = env['account.analytic.account']
+        contracts = m_contract.search([('type', '=', 'contract')])
+        if contracts:
+            _logger.info('Setting payment mode: %d contracts' %
+                         len(contracts))
+        for contract in contracts:
+            payment_mode = contract.partner_id.customer_payment_mode
+            if payment_mode:
```
Comment (yajo) [inline contract_payment_mode/hooks.py:25]: Shouldn't this be `if payment_mode and not contract.payment_mode_id:`?

Comment (pedrobaeza) [inline contract_payment_mode/hooks.py:25]: It can't be, because this module adds the field payment_mode_id.

Comment (yajo) [inline contract_payment_mode/hooks.py:25]: Oops

**Comment by yajo** [comment]
:+1:

**Comment by pedrobaeza** [comment]
Coveralls is dropping a lot because of the UNIT TEST in Travis: https://github.com/OCA/contract/blob/8.0/.travis.yml#L35. I suggest to remove it. Another option is to test the method itself (which by other way, is not a bad idea), that is create a payment mode, a partner with the payment mode, a contract without payment mode, call that method, and test if the contract now have payment mode.

**Comment by yajo** [comment]
Good idea!

**Comment by rafaelbn** [comment]
After https://github.com/OCA/contract/pull/22#issuecomment-176638727 for me is :+1:

**Comment by pedrobaeza** [comment]
Any news on this?

**Comment by pedrobaeza** [comment]
@antespi, can you try to make the test and we merge this?

**Comment by antespi** [comment]
Test added and coveralls is green :smile:

File: contract_payment_mode/tests/test_contract_payment_init.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            'name': 'Test contract',
+            'type': 'contract',
+            'partner_id': self.partner.id,
+            'payment_mode_id': self.payment_mode.id,
+        })
+        self.assertEqual(self._contract_payment_mode_id(contract.id),
+                         self.payment_mode.id)
+
+        contract.payment_mode_id = False
+        self.assertEqual(self._contract_payment_mode_id(contract.id), False)
```
Comment (pedrobaeza) [inline contract_payment_mode/tests/test_contract_payment_init.py:36]: You can do here `self.assertFalse()` instead, but it's not so important

**Comment by pedrobaeza** [comment]
:+1: 

Thanks. Merging.

**Pull request merged**